hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/db/Proto3Codec.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.hadoop.hdds.utils.db;
+
+import org.apache.ratis.thirdparty.com.google.protobuf.CodedOutputStream;
+import org.apache.ratis.thirdparty.com.google.protobuf.InvalidProtocolBufferException;
+import org.apache.ratis.thirdparty.com.google.protobuf.MessageLite;
+import org.apache.ratis.thirdparty.com.google.protobuf.Parser;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.IntFunction;
+
+/**
+ * Codecs to serialize/deserialize Protobuf v3 messages.
+ */
+public final class Proto3Codec<M extends MessageLite>
+    implements Codec<M> {
+  private static final ConcurrentMap<Class<? extends MessageLite>,
+                                     Codec<? extends MessageLite>> CODECS
+      = new ConcurrentHashMap<>();
+
+  /**
+   * @return the {@link Codec} for the given class.
+   */
+  public static <T extends MessageLite> Codec<T> get(Class<T> clazz) {
+    final Codec<?> codec = CODECS.computeIfAbsent(clazz, Proto3Codec::new);
+    return (Codec<T>) codec;
+  }
```

Comment:
What is the benefit of saving codec instances to the map?  Are we likely to have multiple uses of the same proto message in different domain objects?

Reply:
The map is to enforce singleton.  We may consider `Proto3Codec::new` expensive since it uses reflection to create an object.  Suppose we return a new object as below
```java
  public static <T extends MessageLite> Codec<T> getNew(Class<T> clazz) {
    return new Proto3Codec(clazz);
  }
```
Then, it can possibly be misused as
```java
getNew().toCodecBuffer(message, allocator);
```
i.e. it creates a new codec for each serialization.